Object Manager Plus

DartVanya opened the pull request:
- [Implement properties support for Object Manager](https://github.com/winsiderss/systeminformer/commit/2463e2c94dff12e4a647f45763f76cbfe2c735e3)
   - Standard System Informer handle properties window for Windows objects (including Directories).
   - Add support for missing object types: Device, EventPair, Timer, Semaphore, FilterConnectionPort
   - **[System Informer]** Add target link info for Symbolic links
   - Fix Directory handles leaking
   - **[ET Plugin]** Add Windows Object tab to properties window: object attributes and creation time.

- [Enhanced Object Manager](https://github.com/winsiderss/systeminformer/commit/6df00622e3340d3e32f52347f70cb33e9f827257)
   - **[System Informer]** Add GeneralCallbackHandlePropertiesWindowPreOpen, calls on WM_INITDIALOG and allows plugins to customize general handle properties tab.
   - Display real handles count and hide irrelevant access entry. Show object address
   - **Add Symbolic Link Target column**
   - Enhanced navigation to symlink: focus to target in ListView.
- [Defied WinObj: search, statusbar, refresh](https://github.com/winsiderss/systeminformer/commit/535512fbc88cc4c438c2b6d83b45ace8aaf42911)
   - **Implements search using standard SI searchbox**: searches all 3 columns (name, type, target)
   - Status bar (pseudo) with fullpath to selected object (can copy on righclick menu), current directory object count
   - Refresh button - full rebuild of tree and current directory list

-  **[System Informer, Plugins]** **Fixed WM_KEYDOWN forwarding** (both for PH and plugins), broken by https://github.com/winsiderss/systeminformer/commit/5bf7f293067fa20c10427500fca2d2db7c517efe#diff-22c39c7de8c6ce2547b9cda41005993a982b12b5f4d567a14ce7e38534c2bf77L1789
   - [Fixed Ctrl+K to search box in PE VIew](https://github.com/DartVanya/systeminformer/commit/523ee4301c87c4f02997ff909180bfe03123c5c8)
 
- [Add target resolving for Device and ALPC port](https://github.com/DartVanya/systeminformer/commit/c6443fe96a7b4b23361f8eec017f625508421c93)
   - Renamed Symbolic Link Target column to Target
   - Implemented (experimental) asynchronous resolver for Device and ALPC port targets.
      - Resolving Device driver target. Added menu Go to device driver
      - Resolving ALPC server process target. Added menu Go to process...

- [Add target resolving for Mutant, fix search](https://github.com/DartVanya/systeminformer/commit/8dbe1a76743fc0daeb072e1c64a207479beb8b0a)
   - Improve resolver. Added mutant client owner to target and menu Go to thread...
   - Improve filter: it now stay when changing directory. If selected entry match it stay selected and visible when performing search.
   - Tri-state column sort now saved and loaded from settings (ObjectManagerWindowListSort). Sort now works correctly on directory or filter change.
   - Enter - open properties, Ctrl+Enter - open security, Shift+Enter or Shift+Dblclk on symlink opens its properties.
 - [Support for open symlink target in explorer](https://github.com/winsiderss/systeminformer/pull/2197/commits/e4307229521db5a40786cada464b57d82de34735)
 - [Show Job process list in target column](https://github.com/DartVanya/systeminformer/commit/2e873d62b29a3f72c1e6facd175d3f313c92465a)
 - [Add CpuPartition type support and icons for more types](https://github.com/DartVanya/systeminformer/commit/d241e11c2ff46eddaace0a60e93904f49c1a3ae6)
   - Add icon for Job, Semaphore, FilterConnectionPort, CpuPartition, Partition
 
## [New Object Handles page in properties](https://github.com/DartVanya/systeminformer/commit/4707603b5c8bf820d04103b3ef1ad9d56a7a85d5)
   - Old Windows Object page entries moved to general page in "Basic information" block. Windows Object replaced with Handles page.
   - Menu - Go to process... supported
   - Now opens real objects for ALPC Port, FilterConnectionPort, Key (\REGISTRY) using new method (credits to https://github.com/zodiacon/ObjectExplorer)
   - Show additional handles for ALPC Port, Device, Key (match by object name). Extra entries will be highlight with ColorInheritHandles)

## [phlib, kphlib] [Add PhOpenDevice. And Oject Manager Plus FINAL](https://github.com/DartVanya/systeminformer/commit/85b0bcafc93c857ed56d46848bb9160f586388b4)
   - _My first attempt in kernel programming_. Add PhOpenDevice -> KphOpenDevice. Opens Device object handle and optional Device Driver handle. **Can open lowest or topmost device from stack**.
   - Propertines - Device driver info: show topmost and lowest drivers in stack
   - [Theme General] ListVew: fix text readability for hot and selected colored items.
TreeNew:  significantly improve visibility of selected item.

### [Support for multiple non-modal object properties windows](https://github.com/DartVanya/systeminformer/commit/d1fc7bcbe73571e1be5a4329cf55632d6a60309f)
   - Add GeneralCallbackHandlePropertiesUninitializing. Save and load properties window position to/from ObjectManagerPropertiesWindowPosition setting
   - Resolving of Driver Image into Target column, menu - Open file location. Show Driver Image in properties.

### [Add KphOpenObjectByTypeIndex](https://github.com/DartVanya/systeminformer/commit/2f12ae52369bad94426dc9de48b56dcd80bdb74a)
   - [Experimental]. Can open any object types like Callback, Type, etc.
   - New icons for Callback, KeyedEvent, Type, Semaphore

jxy-s commented:
Awesome work! I'll help give an in-depth review when I have some cycles. Some initial feedback on the video you provided (thanks for that it's very helpful). I wonder if it makes sense to change the "Symbolic Link Target" column to a more generic "Target" column to show information/names of things more broadly? I'm thinking, for example, you could show the name for named object or the device for filter ports. Instead of only showing symbol link names.

DartVanya commented:
> Awesome work! I'll help give an in-depth review when I have some cycles. Some initial feedback on the video you provided (thanks for that it's very helpful). I wonder if it makes sense to change the "Symbolic Link Target" column to a more generic "Target" column to show information/names of things more broadly? I'm thinking, for example, you could show the name for named object or the device for filter ports. Instead of only showing symbol link names.

[Add target resolving for Device and ALPC port](https://github.com/DartVanya/systeminformer/commit/c6443fe96a7b4b23361f8eec017f625508421c93)
- Renamed Symbolic Link Target column to Target
- Implemented (experimental) asynchronous resolver for Device and ALPC port targets.
   - Resolving Device driver target. Added menu Go to device driver
   - Resolving ALPC server process target. Added menu Go to server process
- Rewritten navigaton to symbolic links (and device drivers)

https://github.com/user-attachments/assets/ef3729df-dcce-4031-aff6-39576b36bec6

jxy-s commented:
@DartVanya there has been a lot of churn on the PR here. I have been holding off on reviewing since I'm not sure if it's ready for reviews - is it now ready for reviews?

DartVanya commented:
> is it now ready for reviews

Yes it ready for review. I wrote everything I wanted. Last weeks I only using it for myselfs, and seems like all works fine.
PR description is kinda confusing and need to be rewritten with normal changelog and actual video preview.
I synced code to upstream and fixed build. SI, Plugins and KSI should build and run (tested).

Please note, theme fixes PR #2231 also merged in this PR.

jxy-s commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +657,1548 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+    if (PhEqualStringRef(&Context->CurrentPath->sr, &EtObjectManagerRootDirectoryObject, TRUE))
+        fullName = PhFormatString(L"%s%s", PhGetString(Context->CurrentPath), PhGetString(Context->Object->Name));
+    else
+        fullName = PhFormatString(L"%s\\%s", PhGetString(Context->CurrentPath), PhGetString(Context->Object->Name));
```
nit: There are a number of locations that use `PhFormatString` that could be optimized with `PhFormat`.

DartVanya replied:
Replaced with EtGetObjectFullPath

jxy-s commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +657,1548 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+
+    PSYSTEM_HANDLE_INFORMATION_EX handles;
+
+    if (NT_SUCCESS(PhEnumHandlesEx(&handles)))
```
This is both an expensive and invasive operation to try to duplicate a handle. I think we could find a more appropriate solution. Overall we're mostly dealing with "ALPC Port" and "FilterConnectionPort". I'm not sure why "Key" would require this path?

Maybe we should drop the ALPC and communication port parts until we come up with a nicer solution that isn't so invasive to the rest of the system?

jxy-s reviewed (changes requested):
There are changes here to support the implementation that should be broken out. As you already mentioned your other PR (https://github.com/winsiderss/systeminformer/pull/2231) is a prerequisite for this - and I think that should go in before a deeper review. The changes are hard to follow since they're intermixed with the other PR and other supporting changes.

Beyond the other PR the following should be done independently and ahead of this PR:
1. Organize the filter communication port wrappers into `phlib` to be consumed in a more regular manner. Although I do not believe we should be attempting to connect to the communication ports just to extract information. There might be a cleaner way that is less invasive. I don't mind cleaning this up myself, those should probably be moved into `phlib` anyway.
2. The driver changes, if any are necessary, should be done independently of this PR in support of it. I can not accept the changes as they are - the implementation is errant. And due to the nature and extent of your requirements I would prefer to extend the driver APIs myself. I see the types of APIs you would like to have. Allow me some time to reason about them and provide an appropriate (or alternate) implementation for you.

In this PR there are driver changes that break the API contract with user mode. It can not be merged as-is or else builds will fail to work with the existing driver.

I recommend pairing down the changes to remove dependencies on driver changes and connecting to ALPC or filter communication ports. I think there is a lot of great improvements here that can be done and merged first. Then following up with another chunk of work which extends the implementation to fill these gaps - once appropriate supporting code is organized and implemented.

DartVanya commented on KSystemInformer/object.c:
```diff
@@ -2372,6 +2372,41 @@ NTSTATUS KphOpenNamedObject(
+    return KphOpenNamedObject(ObjectHandle,
+                              DesiredAccess,
+                              ObjectAttributes,
+                              KphDynObTypeIndexTable[TypeIndex],
```
I thought that it enough to validate parameters in native.c
Is it correct to check this way from the kernel?
```
    if (AccessMode != KernelMode)
    {
        OBJECT_TYPES_INFORMATION typesInfo = { 0 };

        if (ZwQueryObject(NULL,
            ObjectTypesInformation,
            &typesInfo,
            sizeof(typesInfo),
            NULL) == STATUS_INFO_LENGTH_MISMATCH)
        {
            if (TypeIndex < 2 || TypeIndex > typesInfo.NumberOfTypes)
            {
                return STATUS_INVALID_PARAMETER;
            }
        }
        else
        {
            return STATUS_UNSUCCESSFUL;
        }
    }
```

jxy-s replied:
That might be fine. Although I'll have to check some internals to be certain. Ideally, since we would be using some dyndata anyway - we would extract both the table and the count to avoid the syscall. But it sounds like `KphOpenObjectByTypeIndex` isn't strictly necessary for the changes anyway. I have an idea to be able to provide an implementation without dyndata at all. And might also provide an additional API... I'll get back to you on this.

DartVanya commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -349,57 +503,140 @@ static BOOLEAN NTAPI EtEnumCurrentDirectoryObjectsCallback(
         if (PhEqualStringRef2(TypeName, L"ALPC Port", TRUE))
         {
-            imageIndex = 6;
+            entry->TypeIndex = OBJECT_ALPCPORT;
```
This enumaretion only using internally as listview image index and for target resolver swithing. It never passes to any API functions.
If real type index is needed, code uses PhGetObjectTypeNumberZ to precache or dynamycally obtane index.
Ok, I renamed all occurrences in your recommended manier.

DartVanya commented on phlib/include/kphcomms.h:
```diff
@@ -63,4 +63,17 @@ NTSTATUS KphCommsSendMessage(
     );
 
+_Must_inspect_result_
+__declspec(dllexport)
```
Moved to phlib to native.c

DartVanya commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +657,1548 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+                &portHandle
+            )))
+            {
+                *Handle = portHandle;
```
Changed to STATUS_NOINTERFACE, then we cannot duplicate handle (driver off).
```
        if (!NT_SUCCESS(status = EtpObjectManagerOpenRealObject(Handle, Context, DesiredAccess, PhGetString(Context->Object->TypeName))))
        {
            return STATUS_NOINTERFACE;
        }
```

DartVanya commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +657,1548 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+        {
+            if (NtAlpcConnectPortEx_I)
+            {
+                status = NtAlpcConnectPortEx_I(
```
Changed to STATUS_NOINTERFACE, then we cannot duplicate handle (driver off).

DartVanya commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +657,1548 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
         else
+            deviceName = PhFormatString(L"%s\\%s", PhGetString(Context->CurrentPath), PhGetString(Context->Object->Name));
+
+        if (!NT_SUCCESS(status = PhOpenDevice(Handle, NULL, DesiredAccess, &deviceName->sr, TRUE)))
```
We can extract chunk of usefull information (imho):
- Get info about Lower-edge driver and Upper-edge driver.
   - Menu -  Go to upper/lower device driver
- Device object address.
- Real object attributes, ex. Permanent.

Usage example:

https://github.com/user-attachments/assets/48898563-734e-44e1-a300-4d082402c8c3

DartVanya commented on KSystemInformer/dynimp.c:
```diff
@@ -49,6 +50,46 @@ VOID KphDynamicImport(
+#ifdef _WIN64
+    __try
+    {
+        // ObGetObjectType have equal machine code on 10.0.10240 - 10.0.22631
+        //
+        // nt!ObGetObjectType+0x1C      488d0d[????????]   lea     rcx,[nt!ObTypeIndexTable]
+        // lea  rcx, [rip + off32]
+        PUCHAR rip;
+
+        rip = (PUCHAR)ObGetObjectType + 0x1C;
```
It was experimental implementation, and I know what it is bad and not safe. I wrote this with a purpose to open objects for which public `NtOpenXXX `API isn't existing (like **Type**, Callback). This gives a power to open any object.
Object types table offset is RIP relative to instruction in `ObGetObjectType`. To add it to dyndata we need to know `ObGetObjectType` address, `lea  rcx` instruction address and offset vaule, then we can calculate full address. And we should to it at runtime, since it address points to data section of loaded kernel image (hope I'm not wrong).

It's very interesting, but for now there is no particular need of `KphOpenObjectByTypeIndex`. Type information and statistics can be retrieved without need to open this Type object. Current implementation always open object properties, even if object cannot be opened.

But, if possible, I rally want to keep `PhOpenDevice` (see answer below).

Example, `\ObjectTypes\Token` information with driver and KphOpenObjectByTypeIndex (left), and without (right):

https://github.com/user-attachments/assets/1214bfb4-de3e-4079-a335-7c072b987f47

DartVanya commented:
> As you already mentioned your other PR (#2231) is a prerequisite for this

There is no dependency of theme tweaks here. But I should create separate branch and don't mix all changes in this one.

I did some of requested changes and answered in other conversations. Thank you for appreciating may changes, I hand over the driver code into your hands. :handshake:

jxy-s reviewed (changes requested):
@DartVanya I have implemented for you APIs that will service the needs to open a device. Please see `KphOpenDevice`, `KphOpenDeviceDriver`, and `KphOpenDeviceBaseDevice` here: https://github.com/winsiderss/systeminformer/commit/b82ba21f07e71c1e33ea5b515852f467f0b3c78b

I have not yet implemented the ability to open an arbitrary object by name (using the type index). This will take me more time to investigate and implement and appropriate API for you.

- Please rebase this branch to pick up and use the new device APIs.
- For the time being, please remove `KphOpenObjectByTypeIndex` and the related use in the object manager. After your PR goes in I will continue to investigate an appropriate alternative or solution for this.

Thanks for all the great work here! I'm excited to get this work in <3

DartVanya commented:
> Please see `KphOpenDevice`, `KphOpenDeviceDriver`, and `KphOpenDeviceBaseDevice`

Thanks for good new APIs. I rebased branch 850f2ac. The new APIs work great!
I only changed root directory in `KphOpenDevice` from `NULL` to `capturedAttributes.RootDirectory` because object manager code uses root dir in OA, and with your original code driver returns STATUS_OBJECT_PATH_SYNTAX_BAD

```
        InitializeObjectAttributes(&objectAttributes,
                                   capturedObjectName,
                                   capturedAttributes.Attributes,
                                   capturedAttributes.RootDirectory,
                                   NULL);
```

jxy-s commented:
> I only changed root directory in `KphOpenDevice` from `NULL` to `capturedAttributes.RootDirectory`

Makes sense, tested myself and this should be fine, added here: https://github.com/winsiderss/systeminformer/commit/5f2d516fa53d97624e3ef3b67f6696e323159585

jxy-s commented on phlib/include/phnative.h:
```diff
@@ -1510,6 +1510,18 @@ PhUnloadDriver(
     _In_opt_ PCWSTR Name
     );
 
+PHLIBAPI
+NTSTATUS
+NTAPI
+PhFilterConnectCommunicationPort(
+    _In_ PPH_STRINGREF PortName,
+    _In_ ULONG Options,
+    _In_reads_bytes_opt_(SizeOfContext) PVOID ConnectionContext,
+    _In_ USHORT SizeOfContext,
+    _In_opt_ PSECURITY_ATTRIBUTES SecurityAttributes,
+    _Outptr_ HANDLE* Port
+    );
+
```
I organized the filter APIs into their own file inside of phlib here: https://github.com/winsiderss/systeminformer/commit/38f88f9144e43f2f3aa09a361d4375b5266005d7

This should help clean up your PR a bit 👍

DartVanya replied:
1fae38674a27bfb28d20e72701c5b798e31f0d40

jxy-s reviewed (changes requested):
Left a few more comments that need some attention 👍

jxy-s commented on SystemInformer/mainwnd.c:
```diff
@@ -1363,7 +1347,7 @@ VOID PhMwpOnCommand(
                 L"Desktop",
                 PhpOpenSecurityDesktopHandle,
-                PhpCloseSecurityDesktopHandle,
+                NULL,
```
🤔 is this not going to leak the handle (and below)? Also, how did this ever work? `PhSecurityInformation_Release` passes the `Context` to the `CloseObject` and not the `Handle`? Maybe it was intended for `PhpOpenSecurityStationHandle` to set the output context pointer to the opened handle?

DartVanya replied:
`Handle` is opened from `PhStdGetObjectSecurity` via `OpenObject`. After retrieving of security info by `PhGetObjectSecurity` `Handle` is closing by `NtClose` inside `PhStdGetObjectSecurity`. CloseObject only useful for free resources after security window closes (only Object Manager uses it). Previous `PhpCloseSecurityDesktopHandle` and `PhpCloseSecurityStationHandle` code hadn't effect since `Context` was `NULL`.

If Desktop handle was open by `OpenDesktop`, `NtClose` can't close it and returns `STATUS_HANDLE_NOT_CLOSABLE` (despite handle doesn't have the Protected attribute). But this doesn't happen if handle was duplicated (seems like it's a bug in Windows?). To fix this I added this code to `PhStdGetObjectSecurity`.

```
        if (NtClose(handle) == STATUS_HANDLE_NOT_CLOSABLE &&
            PhEqualString2(this->ObjectType, L"Desktop", TRUE))
        {
            CloseDesktop(handle);
        } 
```

Example of huge Desktop handles leak in current release build:

https://github.com/user-attachments/assets/045f0ae0-2511-4ece-897e-1103ac035a7e

jxy-s replied:
> NtClose can't close it and returns STATUS_HANDLE_NOT_CLOSABLE (despite handle doesn't have the Protected attribute).

That's because it's not in the handle table of the process it's in the win32k subsystem. It's errant to try to close a desktop handle with `NtClose`. They need to go through win32k `NtUserCloseDesktop` (`CloseDesktop`) must be used. Also passing a value to `NtClose` to "try" to close the handle has the possibility of closing the wrong handle. They shouldn't be mixed.

jxy-s replied:
Added a comment elsewhere for a case that wasn't covered. Thanks for the explanation 👍 .

jxy-s commented on tools/peview/main.c:
```diff
@@ -117,13 +117,18 @@ INT WINAPI wWinMain(
+        PvInitializeSettings();
+        PvInitializeSuperclassControls();
+        PhShowWarning2(
+            GetDesktopWindow(),
```
I'm seeing this frequently in the PR, why the changes to pass the desktop window handle in these paths? It seems unnecessary given the documentation and SAL.

DartVanya replied:
Recently @dmex also aked me about this. At your discretion, I can revert this.

> I changed parrent from NULL to GetDesktopWindow() to not show double icon on this dialogs. It wasn't a problem for MessageBox, since it didn't have window icon.
> 
> Parrent `NULL`
> ![image](https://github.com/user-attachments/assets/7ca12402-0eba-4db4-8e9c-d40eee233181)
> Parrent `Desktop`
> ![image](https://github.com/user-attachments/assets/90eb778b-9509-4882-8098-ecdcb590c06a)

jxy-s commented on tools/peview/searchbox.c:
There are edits to `tools/peview/searchbox.c` here, but I don't see any changes to `SystemInformer/searchbox.c`? Unfortunately there is duplicated code somewhat between them. I'm not asking you do consolidate them. But I'm curious (and spot checking) that there are no changes to the complimentary file in `SystemInformer/searchbox.c`.

DartVanya replied:
No, there is no changes in `SystemInformer/searchbox.c`. I only updated code in `tools/peview/searchbox.c` by copy paste whole `SystemInformer/searchbox.c` and renamed `Ph` prefixes to `Pv `.

jxy-s commented on tools/peview/delayhook.c:
```diff
@@ -1213,92 +1405,190 @@ static HWND (WINAPI* PhDefaultCreateWindowEx)(
+    _In_ int            iStateId,
+    _In_ LPCRECT        pRect,
+    _In_ const DTBGOPTS* pOptions
+)
```
```suggestion
    )
```

jxy-s commented on tools/peview/delayhook.c:
```diff
@@ -1213,92 +1405,190 @@ static HWND (WINAPI* PhDefaultCreateWindowEx)(
             {
-                if (PartId == MENU_POPUPGUTTER || PartId == MENU_POPUPBORDERS)
+                // In Windows 11, buttons lack background, making them indistinguishable on dark backgrounds.
+                // To address this, we invert the button. This technique isn't applicable to Windows 10 as it causes the button's border to appear chipped.
```
Just checking, this comment indicates that you intend to only do this on Windows 11 - but I don't see any restrictions to only Windows 11?

DartVanya replied:
V

jxy-s commented on tools/peview/delayhook.c:
```diff
@@ -1448,33 +1732,348 @@ BOOL WINAPI PhSystemParametersInfoHook(
+BOOLEAN CALLBACK PhInitializeTaskDialogTheme(
+    _In_ HWND WindowHandle,
+    _In_opt_ PVOID CallbackData
+)
```
```suggestion
    )
```

jxy-s commented on tools/peview/delayhook.c:
```diff
@@ -1448,33 +1732,348 @@ BOOL WINAPI PhSystemParametersInfoHook(
+        0x1000,
+        PhInitializeTaskDialogTheme,
+        NULL
+    );
```
```suggestion
        );
```

jxy-s commented on tools/peview/delayhook.c:
There is commented out code and comments throughout this file indicating intention to restrict certain changes depending on the operating system version. Spot checking that this is intended as it is?

DartVanya replied:
V

jxy-s commented on SystemInformer/delayhook.c:
```diff
@@ -1230,62 +1457,122 @@ HRESULT PhDrawThemeBackgroundHook(
-                if (PartId == MENU_POPUPGUTTER || PartId == MENU_POPUPBORDERS)
+                // In Windows 11, buttons lack background, making them indistinguishable on dark backgrounds.
+                // To address this, we invert the button. This technique isn't applicable to Windows 10 as it causes the button's border to appear chipped.
+                static enum { yes, no, unknown } mustInvertButton = unknown;
```
I see a comment indicating intention to restrict this to Windows 10, but not runtime check?

jxy-s commented on SystemInformer/delayhook.c:
There is commented out code and comments throughout this file indicating intention to restrict certain changes depending on the operating system version. Spot checking that this is intended as it is?

DartVanya replied:
PhGetThemeClass and its underlying `uxtheme\GetThemeClass` are only available since Win11. I tried to find some workaround, but I couldn't make a correct implementation which will support TaskDialog and doesn't break other UI parts on Windows 10. It's very frustrating why Microsoft uses same theme PartId integers for different control types :confused:

In general, TaskDialog theme currently unsupported in `WindowsVersion < WINDOWS_11`. Thats why I comment out this code, may be someone find solution for Win10. Currently I think what it impossible to do without PhGetThemeClass.

jxy-s commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -1207,57 +2781,160 @@ INT_PTR CALLBACK WinObjDlgProc(
+                        POBJECT_ITEM directory;
+                        ET_OBJECT_ENTRY entry = { 0 };
+
+                        PhGetTreeViewItemParam(context->TreeViewHandle, context->SelectedTreeItem, &directory, NULL);
```
This produces a compilation error when building for release:
```
[ERROR] (1) plugins\ExtendedTools\objmgr.c(2788): error C4703: potentially uninitialized local pointer variable 'directory' used [C:\Users\jxy\source\repos\winsiderss\systeminformer\plugins\ExtendedTools\ExtendedTools.vcxproj]LINK : fatal error LNK1257: code generation failed [C:\Users\jxy\source\repos\winsiderss\systeminformer\plugins\ExtendedTools\ExtendedTools.vcxproj]
```

Needs to be this:
```suggestion
                        if (PhGetTreeViewItemParam(context->TreeViewHandle, context->SelectedTreeItem, &directory, NULL))
                        {
```

jxy-s commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +665,1576 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
-    {
-        DIRECTORY_ENUM_CONTEXT enumContext;
+            if (NT_SUCCESS(status))
+            {
+                DIRECTORY_ENUM_CONTEXT enumContext;
 
-        enumContext.TreeViewHandle = TreeViewHandle;
-        enumContext.RootTreeItem = RootTreeItem;
-        enumContext.DirectoryPath = DirectoryPath;
+                enumContext.TreeViewHandle = TreeViewHandle;
+                enumContext.RootTreeItem = RootTreeItem;
+                enumContext.DirectoryPath = DirectoryPath;
 
-        status = PhEnumDirectoryObjects(
-            directoryHandle,
-            EtEnumDirectoryObjectsCallback,
-            &enumContext
-            );
+                status = PhEnumDirectoryObjects(
+                    directoryHandle,
+                    EtEnumDirectoryObjectsCallback,
+                    &enumContext
+                );
 
-        NtClose(directoryHandle);
-    }
+                NtClose(directoryHandle);
+            }
 
-    TreeView_SortChildren(TreeViewHandle, RootTreeItem, TRUE);
+            TreeView_SortChildren(TreeViewHandle, RootTreeItem, TRUE);
 
-    return status;
+            return status;
```
nit: indentation got screwed up here

jxy-s reviewed (changes requested):
A few more things @DartVanya. I think we're very close to being able to merge 🚀 .

There is some formatting and style nits that I opted not to comment on. Mostly function close parenthesis is not correctly indented or mics style things here and there. I don't mind going to clean those up myself after merging. 👍

jxy-s commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +665,1576 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+                {
+                    KPH_ALPC_COMMUNICATION_INFORMATION connectionInfo;
+
+                    if (NT_SUCCESS(status = KphAlpcQueryInformation(
```
Oddity here - this needs a check for `KsiLevel()` else an assert will trigger on the uninitialized free list. I want to eventually make it so the initialization of the driver isn't a requirement for trying to call the APIs, but that's not how it works today. Please add a check if `KsiLevel() >= KphLevelMed` before calling `KphAlpcQueryInformation`.

DartVanya replied:
Alpc port and driver target resolving only enabled with `BOOLEAN useKsi = KsiLevel() >= KphLevelMed;`

```
        else if (entry->EtObjectType == EtObjectDevice ||       // allow resolving without driver for PhGetPnPDeviceName()
            entry->EtObjectType == EtObjectAlpcPort && useKsi ||
            entry->EtObjectType == EtObjectMutant ||
            entry->EtObjectType == EtObjectJob ||
            entry->EtObjectType == EtObjectDriver && useKsi)
        {
            entry->TargetIsResolving = TRUE;
        }
```

In this case, resolver code will not be queued.
```
    for (ULONG i = 0; i < Context->CurrentDirectoryList->Count; i++)
    {
        ...
        if (!entry->TargetIsResolving)
            continue;
        ...
    }
```

jxy-s commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -420,393 +665,1576 @@ NTSTATUS EtTreeViewEnumDirectoryObjects(
+                HANDLE deviceObject;
+                HANDLE deviceBaseObject;
+
+                if (NT_SUCCESS(status = KphOpenDevice(&deviceObject, DesiredAccess, &objectAttributes)))
```
Similarly here, needs a check for `KsiLevel() >= KphLevelMax`.

jxy-s commented on plugins/ExtendedTools/objprp.c:
```diff
@@ -51,28 +112,645 @@ VOID EtHandlePropertiesInitializing(
+            PhStringRefToUnicodeString(&context->HandleItem->ObjectName->sr, &objectName);
+            InitializeObjectAttributes(&objectAttributes, &objectName, OBJ_CASE_INSENSITIVE, NULL, NULL);
+
+            if (NT_SUCCESS(KphOpenDevice(&deviceObject, READ_CONTROL, &objectAttributes)))
```
And in this path, `KsiLeve() == KphLevelMax`.

jxy-s commented on phlib/secedit.c:
```diff
@@ -1458,7 +1458,11 @@ NTSTATUS PhStdGetObjectSecurity(
     {
         status = PhGetObjectSecurity(handle, SecurityInformation, SecurityDescriptor);
-        NtClose(handle);
+        // NtClose doesn't work for Desktop, this was causing a huge leak of Desktop handles (Dart Vanya)
+        if (!PhEqualString2(this->ObjectType, L"Desktop", TRUE))
+            NtClose(handle);
+        else
+            CloseDesktop(handle);
```
```suggestion
        if (PhEqualString2(this->ObjectType, L"Desktop", TRUE))
            CloseDesktop(handle);
        else if (PhEqualString2(this->ObjectType, L"WindowStation", TRUE))
            CloseWindowStation(handle);
        else
            NtClose(handle);
```

jxy-s commented on phlib/secedit.c:
```diff
@@ -1610,7 +1614,10 @@ NTSTATUS PhStdSetObjectSecurity(
     {
         status = PhSetObjectSecurity(handle, SecurityInformation, SecurityDescriptor);
-        NtClose(handle);
+        if (!PhEqualString2(this->ObjectType, L"Desktop", TRUE))
+            NtClose(handle);
+        else
+            CloseDesktop(handle);
```
```suggestion
        if (PhEqualString2(this->ObjectType, L"Desktop", TRUE))
            CloseDesktop(handle);
        else if (PhEqualString2(this->ObjectType, L"WindowStation", TRUE))
            CloseWindowStation(handle);
        else
            NtClose(handle);
```

jxy-s commented on phlib/theme.c:
```diff
@@ -799,6 +816,7 @@ BOOLEAN CALLBACK PhpThemeWindowEnumChildWindows(
             //case 1: // Old colors
             PhWindowThemeSetDarkMode(tooltipWindow, TRUE);
             PhWindowThemeSetDarkMode(WindowHandle, TRUE);
+            AllowDarkModeForWindow_I(WindowHandle, TRUE);
```
This path is crashing for me on 1809. `AllowDarkModeForWindow_I` is `NULL`.

DartVanya replied:
Moved to phlib, see `PhAllowDarkModeForWindow`, `PhIsDarkModeAllowedForWindow`.

jxy-s replied:
Go ahead and commit this change then I think we can merge 👍

DartVanya replied:
Ok, I added some finishing touches to Object Manager and cleanup code [9c589e6](https://github.com/winsiderss/systeminformer/pull/2197/commits/9c589e6b7718eb11820e94f9a3f54996b5545304)

Add support for KernelObject and KernelOnlyAccess attributes. I think this is quite usefull. NtQueryObject only returns OBJ_EXCLUSIVE and OBJ_PERMANENT.
Please check the driver code, hope this implementation is correct :wink:

https://github.com/user-attachments/assets/1397c45d-7818-4ba2-bf40-4fce2fa418ab

alabuzhev commented on plugins/ExtendedTools/objmgr.c:
```diff
@@ -877,35 +2397,53 @@ INT_PTR CALLBACK WinObjDlgProc(
                 EtObjectManagerRootDirectoryObject
                 );
 
-            {
```
Hi, why this block was removed please?

9ef7295's description doesn't explain it and there seem to be no related comments.

It was [added](https://github.com/winsiderss/systeminformer/commit/b84bdc9a9b4b5761757ca5e46764b1021d67e475#diff-dbc6ac39b248480edbffe8a3f3525e4b0dee98edc4fd9ceeb901bf3726e52ab7R880-R889) in b84bdc9a9b4b5761757ca5e46764b1021d67e475, implementing #2074.

DartVanya replied:
I could removed this by mistake since many things were rewritten and added. I will look into and try to revert this and commit to #2284.

DartVanya replied:
fixes in 165f3a55372a629b8fcbce8d620f0191c2e54e25
